From stable-bounces@linux.kernel.org Tue Feb  6 21:57:59 2007
From: Daniel Walker <dwalker@mvista.com>
Date: Tue, 06 Feb 2007 21:56:37 -0800 (PST)
Subject: Fix ATM initcall ordering.
To: stable@kernel.org
Cc: bunk@stusta.de
Message-ID: <20070206.215637.55835926.davem@davemloft.net>

From: Daniel Walker <dwalker@mvista.com>

[ATM]: Fix for crash in adummy_init()

This was reported by Ingo Molnar here,

http://lkml.org/lkml/2006/12/18/119

The problem is that adummy_init() depends on atm_init() , but adummy_init()
is called first.

So I put atm_init() into subsys_initcall which seems appropriate, and it
will still get module_init() if it becomes a module.

Interesting to note that you could crash your system here if you just load
the modules in the wrong order.

Signed-off-by: Daniel Walker <dwalker@mvista.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 net/atm/common.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- linux-2.6.20.1.orig/net/atm/common.c
+++ linux-2.6.20.1/net/atm/common.c
@@ -816,7 +816,8 @@ static void __exit atm_exit(void)
 	proto_unregister(&vcc_proto);
 }
 
-module_init(atm_init);
+subsys_initcall(atm_init);
+
 module_exit(atm_exit);
 
 MODULE_LICENSE("GPL");
